### DISCUSSION TITLE
Tighter integration with templater plugin.

### DISCUSSION THREAD

**Discussion by imed-ghomari** [Taitava/obsidian-shellcommands#217] [Ideas]
I find myself using SC with templater a lot. Mainly because of the added features provided by this plugin (events, better prompt, different outputs...) and templater because of obsidian API.

For now, the only way (I found) to execute a templater script is to bind it to a command, then execute the associated URI provided by the advanced URI plugin.

There are two downsides to this approach : 
- Advanced URI plugin brings obsidian to the foreground (`Every n second event` isn't usable)
- having to create a command for every template (not ideal)

As I'm sure, many people will use the 2 plugins together, a way to execute these scripts without using advanced URI plugin would be very helpful. I'm thinking, maybe by providing the template path directly in the shell command.

Thank you.

**Comment by Taitava** [comment]
Hi,

thank you for you suggestion! 🙂 I'm really interested in this idea. The Templater plugin is not very familiar to me. I've tried it shortly but haven't really used time to learn it well.

I will need to study Templater more and then I can plan this integration further. But let's start with a quick question: Do you think that it would make sense to create some kind of new [output channel](https://publish.obsidian.md/shellcommands/Output+handling/Output+channels) that calls Templater? I.e. a shell command output would define which templater script to call? This would make it possible to dynamically decide what Templater script to use, if it provides any benefits.

Let's keep planning so we'll see what will come out of this! 😎👍

**Reply by imed-ghomari** [reply]
Yeah, I think an output channel to choose a templater script would be ideal.
Another related feature that came to mind is the ability to pass the variables generated by SC to the templater script (for now, I'm using the clipboard to do that), what do you think ?

**Reply by Taitava** [reply]
> Another related feature that came to mind is the ability to pass the variables generated by SC to the templater script (for now, I'm using the clipboard to do that), what do you think ?

Yes, something like that could be planned. I guess one way to implement it could be via a more generic API for SC (I created a new discussion for it in #219). In particular, calling [`sc.getVariableValue()` or `sc.parseVariables()`](https://github.com/Taitava/obsidian-shellcommands/discussions/219#discussioncomment-2712094) could be done in Templater scripts, I guess.

**Reply by imed-ghomari** [reply]
I noticed that quickadd integration with templater generated a lot of bug request, some of which are :
- using templater and templates together (doesn't work for some reason)
- uncorrect template path in templater settings
- uncorrect templater script path in quickadd settings (SC in this case)

So in order to avoid these types of errors, I would suggest :
- if templates is active and if it conflicts with this, display error prompt
- if templater path is not set, display error prompt
- autocomplete templater script path in SC to avoid typos

**Reply by Taitava** [reply]
Good to know!

> * if templates is active and if it conflicts with this, display error prompt

I'm not yet sure if it's possible to detect plugin conflicts. I'd need to know the conflicts better to know if they are detectable or avoidable. Time will tell. 🙂 


> * if templater path is not set, display error prompt

Will need to look into this, too. I hope Templater can provide some kind of method for checking that its settings are correctly set up. Currently, I don't even know what _Templater path_ is, so I'm not familiar with possible problems it might create. Does it define a folder where Templater looks for script files?

> * autocomplete templater script path in SC to avoid typos

Yea, could be a setting in SC: _Include Templater script path in autocomplete_, only visible in the settings if Templater is installed. Another way to do it manually by users would be to add it to `.obsidian/plugins/obsidian-shellcommands/autocomplete.yaml`: https://publish.obsidian.md/shellcommands/Variables/Autocomplete/Autocomplete#Not+just+for+variables+-+add+custom+suggestions

Yet another way would be to create a variable for it, e.g. `{{templater_script_path}}`.

**Reply by imed-ghomari** [reply]
> Does it define a folder where Templater looks for script files?

Yes, that's it.
